File: src/main/java/work/lclpnet/ap2/game/anvil_fall/AnvilFallSetup.java
package work.lclpnet.ap2.game.anvil_fall;

import net.minecraft.block.BlockState;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockView;
import work.lclpnet.ap2.impl.util.BlockBox;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class AnvilFallSetup {

    private final Pos[] positions;
    private final Random random;

    private AnvilFallSetup(Pos[] positions, Random random) {
        this.positions = positions;
        this.random = random;
    }

    public BlockPos getRandomPosition() {
        Pos randomPos = positions[random.nextInt(positions.length)];
        int offset = random.nextInt(Math.max(randomPos.height, 1));
        return randomPos.pos.up(offset);
    }

    public static AnvilFallSetup scanWorld(BlockView world, BlockBox box, Random random) {
        BlockPos min = box.getMin(), max = box.getMax();

        final int xMin = min.getX(), yMin = min.getY(), zMin = min.getZ();
        final int xMax = max.getX(), yMax = max.getY(), zMax = max.getZ();

        BlockPos.Mutable pos = new BlockPos.Mutable();

        List<Pos> positions = new ArrayList<>();

        for (int x = xMin; x <= xMax; x++) {
            for (int z = zMin; z <= zMax; z++) {
                int height = 0;

                for (int y = yMin; y <= yMax; y++) {
                    pos.set(x, y, z);

                    BlockState state = world.getBlockState(pos);
                    if (!state.getCollisionShape(world, pos).isEmpty()) break;

                    height++;
                }

                if (height <= 0) continue;

                positions.add(new Pos(new BlockPos(x, yMin, z), height));
            }
        }

        return new AnvilFallSetup(positions.toArray(Pos[]::new), random);
    }

    private record Pos(BlockPos pos, int height) {}
}

File: src/minigames/anvil_fall/src/main/kotlin/work/lclpnet/ap2/game/anvil_fall/AnvilFallSetup.kt
package work.lclpnet.ap2.game.anvil_fall

import net.minecraft.core.BlockPos
import net.minecraft.world.level.BlockGetter
import work.lclpnet.gaco.ds.BlockBox
import java.util.*
import kotlin.math.roundToInt

class AnvilFallSetup(
    private val startX: Int,
    private val startY: Int,
    private val startZ: Int,
    private val width: Int,
    private val length: Int,
    private val height: ByteArray,
    private val random: Random,
) {
    private fun getIndexX(x: Int) = x - startX
    private fun getIndexZ(z: Int) = z - startZ
    private fun getIndexFromIndices(ix: Int, iz: Int) = iz * width + ix
    private fun getIndex(x: Int, z: Int) = getIndexFromIndices(getIndexX(x), getIndexZ(z))
    private fun getRandomY(i: Int) = startY + random.nextInt(height[i].toInt())

    fun getRandomPosition(targetX: Int, targetZ: Int, range: Double): BlockPos {
        val x = random.nextGaussian(targetX.toDouble(), range).roundToInt()
        val z = random.nextGaussian(targetZ.toDouble(), range).roundToInt()

        if (x < startX || x >= startX + width || z < startZ || z >= startZ + length) {
            return getRandomPosition()
        }

        return BlockPos(x, getRandomY(getIndex(x, z)), z)
    }

    fun getRandomPosition(): BlockPos {
        var i: Int
        var j = 0

        do {
            i = random.nextInt(height.size)
        } while (height[i] <= 0 && j++ < 16)

        val x: Int
        val y: Int
        val z: Int

        if (height[i] <= 0) {
            x = (2 * startX + width - 1) / 2
            y = startY
            z = (2 * startZ + length - 1) / 2
        } else {
            x = startX + i % width
            y = getRandomY(i)
            z = startZ + i / width
        }

        return BlockPos(x, y, z)
    }

    fun getRandomPositionAt(x: Int, z: Int): BlockPos {
        val i = getIndex(x, z)

        if (i < 0 || i >= height.size || height[i] <= 0) {
            return getRandomPosition(x, z, 2.0)
        }

        return BlockPos(x, getRandomY(i), z)
    }

    companion object {
        fun scanWorld(world: BlockGetter, box: BlockBox, random: Random): AnvilFallSetup {
            val min = box.min()
            val max = box.max()

            val xMin = min.x; val yMin = min.y; val zMin = min.z
            val xMax = max.x; val yMax = max.y; val zMax = max.z
            val width = xMax - xMin + 1
            val length = zMax - zMin + 1

            val pos = BlockPos.MutableBlockPos()
            val heights = ByteArray(width * length)

            for (x in xMin..xMax) {
                for (z in zMin..zMax) {
                    var heightCount = 0

                    var y = yMin
                    while (y <= yMax && heightCount < Byte.MAX_VALUE) {
                        pos.set(x, y, z)
                        val state = world.getBlockState(pos)
                        if (!state.getCollisionShape(world, pos).isEmpty) break
                        heightCount++
                        y++
                    }

                    if (heightCount < 0) heightCount = 0

                    val ix = x - xMin
                    val iz = z - zMin
                    heights[iz * width + ix] = heightCount.toByte()
                }
            }

            return AnvilFallSetup(xMin, yMin, zMin, width, length, heights, random)
        }
    }
}
